openmc/deplete/batchwise.py

```diff
@@ -0,0 +1,1316 @@
+        if isinstance(val, Material):
+            check_value("Material", str(val.id), self.burn_mats)
+
+        elif isinstance(val, str):
+            if val.isnumeric():
+                check_value("Material id", val, self.burn_mats)
+                val = [mat for mat in self.model.materials if mat.id == int(val)][0]
+
+            else:
+                check_value(
+                    "Material name",
+                    val,
+                    [mat.name for mat in self.model.materials if mat.depletable],
+                )
+                val = [mat for mat in self.model.materials if mat.name == val][0]
+
+        elif isinstance(val, int):
+            check_value("Material id", str(val), self.burn_mats)
+            val = [mat for mat in self.model.materials if mat.id == val][0]
+
+        else:
+            ValueError(f"Material: {val} is not supported")
+
+        return val
+
+    @abstractmethod
+    def _model_builder(self, param):
+        """Builds the parametric model to be passed to the
+        :meth:`openmc.search.search_for_keff` method.
+        Callable function which builds a model according to a passed
+        parameter. This function must return an openmc.model.Model object.
+
+        Parameters
+        ----------
+        param :
+            Model material function variable
+
+        Returns
+        -------
+        _model :  openmc.model.Model
+            Openmc parametric model
+        """
+
+    def search_for_keff(self, x, step_index):
+        """Perform the criticality search on parametric material variable.
+        The :meth:`openmc.search.search_for_keff` solution is then used to
+        calculate the new material volume and update the atoms concentrations.
+
+        Parameters
+        ----------
+        x : list of numpy.ndarray
+            Total atoms concentrations
+
+        Returns
+        -------
+        x : list of numpy.ndarray
+            Updated total atoms concentrations
+        root : float
+             Search_for_keff returned root value
+        """
+        # Update AtomNumber with new conc vectors x. Materials are also updated
+        # even though they are also re-calculated when running the search_for_kef
+        self._update_materials(x)
+
+        # Solve search_for_keff and find new value
+        root = self._search_for_keff(0)
+
+        # Update concentration vector and volumes with new value
+        volumes = self._calculate_volumes(root)
+        x = self._update_x_and_set_volumes(x, volumes)
+
+        return x, root
+
+
+class BatchwiseMaterialRefuel(BatchwiseMaterial):
+    """Batch wise material-based class for refuelling (addition or removal) scheme.
+
+    A user doesn't need to call this class directly.
+    Instead an instance of this class is automatically created by calling
+    :meth:`openmc.deplete.Integrator.add_batchwise` method from an integrator
+    class, such as  :class:`openmc.deplete.CECMIntegrator`.
+
+    .. versionadded:: 0.13.4
```

Comment:
Same comment on the version added string and allowing the user to construct these directly